src/PromiseQueue.ts

```diff
@@ -0,0 +1,64 @@
+type Task = () => Promise<void>
+
+export interface PromiseQueue {
+  enqueue(task: Task): void
+  flushQueue(): Promise<void>
+}
+
+export class SimplePromiseQueue implements PromiseQueue {
+  private readonly _queue: Task[] = []
+  private _runner: Promise<void> | null = null
+
+  public enqueue(task: Task) {
+    this._queue.push(task)
+    this.flushQueue()
+  }
+
+  public async flushQueue(): Promise<void> {
+    const runner = async () => {
+      while (this._queue.length) {
+        const nextTask = this._queue.shift()!
+        await nextTask()
+      }
+      this._runner = null
+    }
+
+    if (this._runner) {
+      return await this._runner
+    }
+    this._runner = runner()
+    await this._runner
+  }
+}
+
+/**
+ * A "queue" that only executes the current task and the most recently
+ * enqueued task. If a new task arrives while there is a task queued up,
+ * the new task replaces the previous task.
+ */
+export class SingleTaskPromiseQueue {
+  private _task: Task | null = null
+  private _runner: Promise<void> | null = null
+
+  public enqueue(task: Task) {
+    this._task = task
+    this.flushQueue()
+  }
+
+  public async flushQueue(): Promise<void> {
+    const runner = async () => {
+      while (this._task) {
+        const task = this._task
+        this._task = null
+        await task()
+      }
+      this._runner = null
+    }
+
+    if (this._runner) {
+      return await this._runner
+    }
+    this._runner = runner()
+    await this._runner
```

Comment:
One option here is to run this in a `setTimeout`. Then if we do a bunch of mutations, we only do a single `saveState` instead of the two here (as implemented, we will call `saveState` for the first mutation, then the last one). I did things this way for simplicity but I don't think the other way is too bad.